README.md
Currentscape is a Pythonic API enabling scientists to easily plot 
currentscapes figures in the style of
[Leandro M. Alonso and Eve Marder, ”Visualization of currents in neural models with similar behavior and different conductance densities” (2019)](https://datadryad.org/stash/dataset/doi:10.5061/dryad.d0779mb)
, given the currents and voltage traces of an in-silico neuron cell.
Currentscape figures plot the percentage of each inward and outward ionic current,
the total inward and outward currents, as well as the voltage in function of time.
This API is useful to see what current plays a major (or a minor) role at a given time,
and see in depth the currents dynamics.

This API is based on the code provided by the susmentioned paper.


![Illustration of a currentscape plot](doc/source/images/plot.png)

### Citation

When you use the Currentscape software or method for your research, we ask you to cite the following publication (this includes poster presentations):

[Leandro M. Alonso and Eve Marder, ”Visualization of currents in neural models with similar behavior and different conductance densities” (2019)](https://elifesciences.org/articles/42722)

    @article {10.7554/eLife.42722,
    article_type = {journal},
    title = {Visualization of currents in neural models with similar behavior and different conductance densities},
    author = {Alonso, Leandro M and Marder, Eve},
    editor = {Westbrook, Gary L and Skinner, Frances K and Lankarany, Milad and Britton, Oliver},
    volume = 8,
    year = 2019,
    month = {jan},
    pub_date = {2019-01-31},
    pages = {e42722},
    citation = {eLife 2019;8:e42722},
    doi = {10.7554/eLife.42722},
    url = {https://doi.org/10.7554/eLife.42722},
    abstract = {Conductance-based models of neural activity produce large amounts of data that can be hard to visualize and interpret. We introduce visualization methods to display the dynamics of the ionic currents and to display the models’ response to perturbations. To visualize the currents’ dynamics, we compute the percent contribution of each current and display them over time using stacked-area plots. The waveform of the membrane potential and the contribution of each current change as the models are perturbed. To represent these changes over a range of the perturbation control parameter, we compute and display the distributions of these waveforms. We illustrate these procedures in six examples of bursting model neurons with similar activity but that differ as much as threefold in their conductance densities. These visualization methods provide heuristic insight into why individual neurons or networks with similar behavior can respond widely differently to perturbations.},
    keywords = {neuronal oscillators, Na+ channels, Ca++ channels, K+ channels, conductance-based, ionic channels},
    journal = {eLife},
    issn = {2050-084X},
    publisher = {eLife Sciences Publications, Ltd},
    }


### Installing currentscape

Currentscape can be pip installed with the following line:

    pip install currentscape

This installs the currentscape module and its dependencies.

If you want to be able to run the currentscape examples, you will need to also install the example dependencies. This can be done with the following line:

    pip install currentscape[example]

### Loading currentscape in Python

After installing currentscape, your PYTHONPATH environment variable should normally
contain the directory where the currentscape module is installed. Loading currentscape
in Python becomes then as easy as:

    import currentscape

### Plotting your first currentscape

Given voltage and current data (see 'Extracting currents and ionic concentrations' section below for how to get voltage and currents from a cell), as well as an adequate config json file, producing a currenscape figure should be as simple as

    import os
    import numpy as np
    from currentscape.currentscape import plot_currentscape

    data_dir = "path/to/current/recording/files"
    currs = [
        "i_pas",
        "ihcn_Ih",
        "ica_Ca_HVA2",
        "ica_Ca_LVAst",
        "ik_SK_E2",
        "ik_SKv3_1",
        "ik_K_Pst",
        "ik_K_Tst",
        "ina_NaTg",
    ]

    # load voltage data
    v_path = os.path.join(data_dir, "v.dat")
    voltage = np.loadtxt(v_path)[:, 1]

    # load currents from files
    currents = []
    for curr in currs:
        file_path = os.path.join(data_dir, f"{curr}.dat")
        currents.append(np.loadtxt(file_path)[:, 1])
    currents = np.array(currents)

    # define config
    config = "path/to/config.json"
    # can also pass config as a dictionnary, as commented below
    # curr_names = ["pas", "Ih", "Ca_HVA2", "Ca_LVAst", "SK_E2", "SKv3_1", "K_Pst", "K_Tst", "NaTg"]
    # config = {
    #     "current": {"names": curr_names},
    #     "legendtextsize": 5,
    # }

    # produce currentscape figure
    fig = plot_currentscape(voltage, currents, config)
    fig.show()

The voltage should be a list of floats corresponding to the voltage at each timestep.

The currents should be a list of lists of floats corresponding to each current at each timestep.

Each current list and the voltage list should have the same size.


### About the config


Here is an example of a config file containing all defaults values :

    {
        "show": {
            "currentscape": true,
            "ylabels": true,
            "yticklabels": true,
            "xlabels": false,
            "xticklabels": false,
            "_comment1": "If enabled, xgridlines plot vertical lines in all plots at xticks positions.",
            "xgridlines": false,
            "legend": true,
            "all_currents": false,
            "_comment2": "total contribution plots two pie charts (positive and negative) showing the contribution of each current over the whole simulation.",
            "total_contribution": false
        },
        "current": {
            "_comment1": "is not set by default.  The current names should appear in the same order as in the currents argument. Is mandatory if ['show']['legend'] is true",
            "names": [
                "Na",
                "CaT",
                "CaS",
                "A",
                "KCa",
                "Kd",
                "H",
                "L"
            ],
            "_comment2": "if True, reorder currents with decreasing order of %.",
            "reorder": true,
            "_comment3": "if True, do not take into account ticks and ylim below.",
            "autoscale_ticks_and_ylim": true,
            "_comment4": "only taken into account if autoscale_ticks_and_ylim is False",
            "ticks": [
                5,
                50,
                500
            ],
            "_comment5": "only taken into account if autoscale_ticks_and_ylim is False",
            "ylim": [
                0.01,
                1500
            ],
            "units": "[pA]",
            "_comment6": "color for summed currents.",
            "color": "black",
            "_comment7": "True to plot absolute currents with stackplots, False to plot them with lines",
            "stackplot": false,
            "_comment8": "thickness of black line separating the inward & outward stackplots. in %age of y size of plot.",
            "black_line_thickness": 2,
            "_comment9": "only used if stackplot is True",
            "legacy_method": false
        },
        "currentscape": {
            "in_label": "inward %",
            "out_label": "outward %",
            "_comment1": "only used when legacy_method is true. if too low, white pixels can appear at the bottom of currentscape plots because of rounding errors. Only used when use_legacy_method is True.",
            "y_resolution": 10000,
            "legacy_method": false
        },
        "ions": {
            "_comment1": "if True, do not take into account ticks and ylim below.",
            "autoscale_ticks_and_ylim": true,
            "_comment2": "only taken into account if autoscale_ticks_and_ylim is False",
            "ticks": [
                0.0005,
                0.05,
                5
            ],
            "_comment3": "only taken into account if autoscale_ticks_and_ylim is False",
            "ylim": [
                0.00001,
                100
            ],
            "units": "[mM]",
            "_comment4": "if True, reorder currents with decreasing order",
            "reorder": true,
            "_comment5": "is not set by default.  The ions concentration names should appear in the same order as in the ions argument. Is mandatory if ['show']['legend'] is true",
            "names": [
                "cai",
                "ki",
                "nai"
            ]
        },
        "colormap": {
            "name": "Set1",
            "_comment1": "color number. Taken into account only if pattern use is True",
            "n_colors": 8
        },
        "stackplot": {
            "_comment1": "only used when ['currentscape']['legacy_method'] is true. data along x axis are summed up into chunks when pattern use is True. Put to 1 to disable.",
            "x_chunksize": 50
        },
        "pattern": {
            "use": false,
            "patterns": ["", "/", "\\", "x", ".", "o", "+"],
            "density": 5,
            "linewidth": 0.2,
            "_comment1": "since the pattern color is defined by the edgecolor, this parameter also changes the edgecolor of the pie charts",
            "color": "black"
        },
        "line": {
            "_comment1": "Is used when ['pattern']['use'] and ['show']['all_currents'] are True and ['current']['stackplot'] is False. Should have the same length as ['pattern']['patterns']",
            "styles": [
                "solid",
                [0, [1, 1]],
                [0, [2, 1]],
                [0, [2, 1, 1, 1]],
                [0, [2, 1, 1, 1, 1, 1]],
                [0, [2, 1, 2, 1, 1, 1]],
                [0, [2, 1, 2, 1, 1, 1, 1, 1]]
            ]
        },
        "voltage": {
            "ylim": [-90, 30],
            "ticks":[-50, -20],
            "units": "[mV]",
            "color": "black",
            "horizontal_lines": true
        },
        "xaxis": {
            "units": "[ms]",
            "_comment1": "if None, xticks are generated automatically. Can put a list of xticks to force custom xticks.",
            "xticks": null,
            "gridline_width": 1,
            "gridline_color": "black",
            "gridline_style": "--"
        },
        "output": {
            "savefig": false,
            "dir": ".",
            "fname": "test_1",
            "extension": "pdf",
            "dpi": 400,
            "transparent": false
        },
        "legend": {
            "textsize": 4,
            "bgcolor": "lightgrey",
            "_comment1": "1. : top of legend is at the same level as top of currentscape plot. higher value put legend higher in figure.",
            "ypos": 1.0,
            "_comment2": "forced to 0 if ['pattern']['use'] is False and ['current']['stackplot'] is False",
            "handlelength": 1.4
        },
        "figsize": [
            3,
            4
        ],
        "title": null,
        "titlesize": 12,
        "labelpad": 1,
        "textsize": 6,
        "lw": 0.5,
        "adjust": {
            "left": 0.15,
            "right": 0.85,
            "top": null,
            "bottom": null
        }
    }

If you do not want to modify the default values, you should at least specify the current names if you want to plot with the legend.
Your configuration file could be as small as:

    {
        "current": {
            "names": [
                "Na",
                "CaT",
                "CaS",
                "A",
                "KCa",
                "Kd",
                "H",
                "L"
            ],
    }

The config argument can be passed as a dictionnary, or as a path to a json file.
As data can vary greatly, it is recommended to adapt the config file consequently.
One may want to change the y axis limits, or the ticks, for example.
If the legend is cut, one may decrease the legendsize, the adjust right parameter or increase the figsize.


### Setting the colormap

Since each color of the colormap applies to one category (one current), using categorical / qualitative colormaps is recommended.
These colormaps have colors chosen to easily distinguish each category.

Also, be careful not to use any colormap that uses white, since white is the default color when there is no data (no inward or outward currents).
It would be then hard to know if there is a 'white' current, or no current at all.
Using a colormap that uses black is also not advised, since the plots on top and bottom of currentscapes, 
as well as the line separating the inward and outward currentscapes, are black. 
If a black current ends up near the top or bottom of the plot, it would decrease readability.

You can set your colormap using "colormap":{"name": "the_name_of_the_colormap"} in the config file.
The name of the colormap can be one of the matplotlib colormaps (https://matplotlib.org/3.1.0/tutorials/colors/colormaps.html), 
or one of the palettable module (https://jiffyclub.github.io/palettable/).
The palettable colormaps should be inputted in the form "origin.palette_N", N being the number of different colors (i.e. the number of currents if patterns are not used.)
Example:
    "cartocolors.qualitative.Safe_8"


### Showing x axis label, ticklabel, gridlines

You can use the configuration to show x axis label, ticklabels and vertical gridlines. 
If you choose to display them, the label and ticklabels will only show on the bottom plot, and the vertical gridlines will show on all plots, and correspond to the x ticks (generated automatically, if not set in the config). 
However, to show ticklabels and gridlines, you have to also input time as an argument to the plot_currentscape function. Here is an example:

    # load voltage data
    data_dir = "path/to/data/dir"
    v_path = os.path.join(data_dir, "v.dat")
    time = np.loadtxt(v_path)[:, 0]
    voltage = np.loadtxt(v_path)[:, 1]

    currents = load_current_fct(data_dir)
    config = "path/to/config.json"

    # produce currentscape figure
    fig = plot_currentscape(voltage, currents, config, time=time)

Be aware that the time data are expected to grow monotonically.

Also, when setting custom x ticks through the config, try to stick with ticks within time data limits for optimal display.

### Using patterns

If you have a lot of currents to display and do not find a colormap with enough colors to distinguish them all, you can use patterns (also called hatches).
Note: if you are using a lot of currents, you may want to increase the "legend":"ypos" (e.g. to 1.5) in your config to have a legend higher in the figure.

By putting "pattern": {"use": True} in your config, currentscape will put patterns like stripes or dots on top of your currents, 
and it will mix colors and patterns so that two successive currents do not have the same pattern or color.
In the "pattern" key of your config, you can increase the 'density' (frequency) or your patterns, change the pattern linewidth, color, etc.
You can also change the patterns or the number of different colors to use with the adequate config.

You could also want to use pattern if you are using a non-qualitative colormap that do not have a lot of distinguishable colors.


### Showing all absolute currents

By putting "show":{"all_currents": True} in the config file, two subplots showing all the positive and negative currents are added at the bottom of the figure.
The currents can be displayed as stackplots by putting "current":{"stackplot": True} in the config, or as lines, by putting "current":{"stackplot": False} in the config. In case they are displayed with lines, while using patterns for the current shares, the lines will be displayed with styles (dashed, dotted, etc.). In such a case, the number of line styles should be equal to the number of patterns (which they are, by default). Keep this in mind when changing either the line styles or the patterns.


### Using legacy methods

You can use currentscape legacy methods by setting "currentscape": {"legacy_method": True} in the config.
If you want to show all currents with a stackplot, you can also use its legacy method by setting "current": {"legacy_method": True} in the config.
The legacy methods can take longer to compute, take more memory during computation and
the legacy barplot method (used when "pattern": {"use": True}, or when both "current": {"stackplot": True} and "show": {"all_currents": True}) has a bad display when the figure is saved in the pdf format.

However, these methods can be useful to display the main features of the plots, without having the details blurred by e.g. low resolution.


### Showing ionic concentrations

You can plot the ionic concentrations in a subplot at the bottom of the figure by passing your ionic concentration data to the main function: plot_currentscape(voltage, currents, config, ions), and by passing the ion names to the config under: "ions":{"names":your_list}. Note that, as for the currents, the ion names should correspond to the ion data (i.e. be listed in the same order).


### Showing overall contribution pie charts

By setting "show":{"total_contribution": True} in the configuration, two pie charts are added at the bottom of the figure, each showing the overall contribution of each current over the whole simulation, one for the outward currents, and the other one for the inward currents.


### Extracting currents and ionic concentrations

You can see an example of how to extract currents and ionic concentractions with bluepyopt and emodelrunner in the example folder: `examples/use_case`.
Please note that you should have bluepyopt, emodelrunner and NEURON installed in order to run the example.
The example folder contains
a cell,
a script to run the cell by applying to it a step stimulus and record its current and ionic concentration traces,
and another script to plot its currentscape.

To run the cell, go to `examples/use_case` and do

    sh run_py.sh

Once this is done, you can plot the currentscape by doing:

    python plot.py

You can adjust the currentscape plot by modifying the configuration that is hard-coded in `plot.py`.


### Known caveats

Since currentscape can deal with a lot of data, it sometimes crashes with an error message such as `Bus error` or `Killed` when it runs out of memory. You can solve this error by allocating more memory. Note that this error is prone to happen more frequently when legacy methods are used.


CHANGELOG.rst
Changelog
=========

Version 0.1.1
-------------

Changes
~~~~~~~
- examples is not be treated as a module anymore (as it should). To use the examples, one should go to the example folder to run the scripts.
- Added line creation in get_colors_and hatches_lists, renamed it get_colors_hatches_lines_lists and used it in more places.

Added
~~~~~
- Example with a cell, script to record currents/ionic concentrations from it, and script to plot its currentscape.
- Test to run the new example, running the cell and plotting its currentscape.

Fixed
~~~~~
- Fixed legend for currentscape with no hatches
- Removed minor yticks. They were popping out when not using autoscale for ylim and ticks with log scale
- Dotted horizontal lines at tick level. They were not the right length when data with adaptable timesteps were given

Removed
~~~~~~~
- extract_currs module
- unused .gitreview file
- __init__.py file in examples. examples folder should not be treated as a module.


Version 0.1.0
-------------

Changes
~~~~~~~
- use fill_between instead of imshow or barplot (more precise, less pixelised and consumes less memory)
- default extension for figure saving in config is now pdf (in order to not loose details)

Added
~~~~~
- new 'legacy_method' entry in config for currents and currentscape to access legacy imshow and barplots


Version 0.0.12
--------------

Changes
~~~~~~~
- Turned some print statements into logging statements
- Refactored Main Protocol's run function in extract_currs
- Refactored output writing into a write_output function in extract_currs
- Turned os.path statements into pathlib.Path statements
- Refactored extract_currs/features/define_efeatures
- Turned string formatting into f-strings
- Refactored the loading of release_params into a function
- Moved bluepyopt, neuron and scipy dependencies as extras in setup.py
- Refactored currents.reorder_currents_and_names for better quality code
- Used np.asarray instead of np.array when  possible
- Refactored color and pattern mapping into functions
- Renamed col_idxs into curr_idxs in several function arguments

Added
~~~~~
- Unit tests for all currentscape functions that do not use matplotlib
- Linting checks to the extract_currs module
- Test for the new write_output function
- Installation instructions in the README

Removed
~~~~~~~
- Unused logging.basicConfig
- Print statements
- extract_currs/YeoJohnson.py
- extract_currs/features/eFELFeatureExtra
- extract_currs/features/SingletonWeightObjective
- Non-necessary stuff from extract_currs/features/define_efeatures
- Score calculations during Main Protocol
- Non-necessary variables form extract_currs' create_protocols
- __init__ function in SweepProtocolCustom that was the same as the parent class

Fixed
~~~~~
- Tox
- Barplot shrinking when chunk size has to be resized


Version 0.0.11
--------------

Changes
~~~~~~~
- Improved memory use by deleting large variables when they are not of use anymore.
- Improved memory use by using np.int8 in create_black_line function.
- Made sure that every data in Datasets are numpy arrays.


Version 0.0.10
--------------

Changes
~~~~~~~
- Forced fixed timesteps in extract_currs, since currentscape expects data with fixed timesteps.

Version 0.0.9
-------------

Added
~~~~~
- New SAHP stimulus for current extraction.



Version 0.0.8
-------------

Fixed
~~~~~
- Test data in test_currentscape_functions.


Version 0.0.7
-------------

Fixed
~~~~~
- Unit conversion when turning current density into currents in extract_currs.

Added
~~~~~
- Added tests for data processing functions.
- Added warning when there is not enough colors in colormap.

Changes
~~~~~~~
- Refactored currentscape into smaller, more meaningful modules.
- Changed default left-side adjusting, 
    so that the labels and ticks do not overlap.
- Changed dict key for blackline thickness and x-chunksize in config
    since they should not belong exclusively to currentscape anymore.
- Changed 'labels' and 'ticklabels' to 'ylabels' and 'yticklabels' in config["show"]
    to distinguish them from 'xlabels' and 'xticklabels'.

New Features
~~~~~~~~~~~~
- Reorder the currents in each subplot, 
    so that the currents with the largest contribution lay at the top.
    The legend is reordered accordingly.
- Added an autoscaling for y limits and ticks for current (and ions) subplots.
- Added a new subplot to currentscape: absolute currents.
- Absolute currents can be displayed as lines or as a stackplot.
- Allowed extract_currs to extract ionic concentration.
- Added a new subplot to currentscape: ionic concentration.
- Added a new subplot: pie charts showing the overall contribution of currents.
- Label and ticklabel of the x axis can now be displayed on the bottom plot.
- Gridlines corresponding to the x ticks can be displayed on all the plots.
- x ticks can be custom or generated automatically.
- Currentscape plotting can be disabled in the config.


Version 0.0.6
-------------

Fixes
~~~~~
- Currentscape used to fill empty data (e.g. no inward/outward current at all)
    with the first color in data. Now it fills it with white.
- Fixed config example in README and doc.
- Removed top and bottom frame of currentscape which was hiding part of the data.
- Fixed central black line separating currentscape plots that weas hiding part of the data.

New Features
~~~~~~~~~~~~
- Added new colormap choices.
- Added possibility of putting patterns (hatches) on top of color palette.
- When using patterns, matplotlib uses bar plots, which takes some time to compute,
    so reducing x-resolution was enabled.
- Chunksize (when reducing x-resolution) is corrected when not a divisor of data size.
- Patterns color, design and 'density' can be customised.
- The legend y position can be customised.
- Added a 'mixer' that prevent two successive currents to have the same color or pattern.
- Added library to easily extract currents from any location (soma or other).

currentscape.py
"""Module for plotting currentscapes as in https://datadryad.org/stash/dataset/doi:10.5061/dryad.d0779mb."""
"""The main function is based on scripts from the susmentioned article
    that are under the CC0 1.0 Universal (CC0 1.0) Public Domain Dedication license."""

from pylab import *
import json
import os


def replace_defaults(config, new_config):
    """Recursive fct. Replace defaults value with user input."""
    for key, item in new_config.items():
        if isinstance(item, dict):
            item = replace_defaults(config[key], item)
        config[key] = item

    return config


def set_default_config(c):
    """Set non-specified values in config to default."""
    if c is None:
        c = {}
    elif isinstance(c, str):
        f = open(c, "r")
        c = json.load(f)

    config = {}
    show = {}
    show["labels"] = True
    show["ticklabels"] = True
    show["legend"] = True
    config["show"] = show

    current = {}
    current["ticks"] = [5, 50, 500]
    current["ylim"] = (0.01, 1500)
    current["units"] = "[pA]"
    current["color"] = "black"
    config["current"] = current

    currentscape = {}
    currentscape["in_label"] = "inward %"
    currentscape["out_label"] = "outward %"
    currentscape["cmap"] = "Set1"
    currentscape["y_resolution"] = 1000
    config["currentscape"] = currentscape

    voltage = {}
    voltage["ticks"] = None
    voltage["ylim"] = (-90, 30)
    voltage["units"] = "[mV]"
    voltage["color"] = "black"
    voltage["horizontal_lines"] = True
    config["voltage"] = voltage

    output = {}
    output["savefig"] = False
    output["dir"] = "."
    output["fname"] = ""
    output["extension"] = "png"
    output["dpi"] = 400
    output["transparent"] = False
    config["output"] = output

    config["title"] = None
    config["figsize"] = (3, 4)
    config["labelpad"] = 1
    config["textsize"] = 6
    config["legendsize"] = 6
    config["titlesize"] = 12

    adjust = {}
    adjust["left"] = None
    adjust["right"] = 0.85
    adjust["top"] = None
    adjust["bottom"] = None
    config["adjust"] = adjust

    return replace_defaults(config, c)


def remove_ticks_and_frame(ax):
    """Remove ticks (but not ytick label) and frame."""
    ax.tick_params(
        axis="x",  # changes apply to the x-axis
        which="both",  # both major and minor ticks are affected
        bottom=False,  # ticks along the bottom edge are off
        top=False,  # ticks along the top edge are off
        labelbottom=False,  # labels along the bottom edge are off
    )
    ax.tick_params(
        axis="y",  # changes apply to the x-axis
        which="both",  # both major and minor ticks are affected
        left=False,  # ticks along the bottom edge are off
        right=False,  # ticks along the top edge are off
        pad=0,
    )
    ax.spines["top"].set_visible(False)
    ax.spines["right"].set_visible(False)
    ax.spines["bottom"].set_visible(False)
    ax.spines["left"].set_visible(False)


def adjust(adjust_left, adjust_right, adjust_top, adjust_bottom):
    """Adjust Subplots."""
    subplots_adjust(wspace=0, hspace=0)
    if adjust_left:
        subplots_adjust(left=adjust_left)
    if adjust_right:
        subplots_adjust(right=adjust_right)
    if adjust_top:
        subplots_adjust(top=adjust_top)
    if adjust_bottom:
        subplots_adjust(bottom=adjust_bottom)


def set_label(ax, x, y, label, textsize):
    """Set text as label on the y axis."""
    ax.text(
        x,
        y,
        label,
        horizontalalignment="right",
        verticalalignment="center",
        rotation="vertical",
        size=textsize,
    )


def set_legend(ax, im, curr_names):
    """Set each current name color-coded in legend."""
    # create a patch (proxy artist) for every current
    patches = [
        matplotlib.patches.Patch(label=curr_names[i]) for i in range(len(curr_names))
    ]
    # put those patched as legend-handles into the legend
    leg = ax.legend(
        handles=patches, bbox_to_anchor=(1.01, 1), loc=2, borderaxespad=0.0,
    )
    # set legend label color
    for i_color, text in enumerate(leg.texts):
        text.set_color(im.cmap(im.norm(i_color)))


def plotCurrentscape(voltage, currents, config):
    """Returns a figure containing current scapes."""
    # load config and set default to unspecified terms
    c = set_default_config(config)
    # make a copy of currents
    # CURRENTSCAPE CALCULATION STARTS HERE.
    curr = array(currents)
    cpos = curr.copy()
    cpos[curr < 0] = 0
    cneg = curr.copy()
    cneg[curr > 0] = 0

    normapos = sum(abs(array(cpos)), axis=0)
    normaneg = sum(abs(array(cneg)), axis=0)
    npPD = normapos
    nnPD = normaneg
    cnorm = curr.copy()
    cnorm[curr > 0] = (abs(curr) / normapos)[curr > 0]
    cnorm[curr < 0] = -(abs(curr) / normaneg)[curr < 0]

    resy = 1000
    impos = zeros((resy, shape(cnorm)[-1]))
    imneg = zeros((resy, shape(cnorm)[-1]))

    times = arange(0, shape(cnorm)[-1])
    for t in times:
        lastpercent = 0
        for numcurr, curr in enumerate(cnorm):
            if curr[t] > 0:
                percent = int(curr[t] * (resy))
                impos[lastpercent : lastpercent + percent, t] = numcurr
                lastpercent = lastpercent + percent
    for t in times:
        lastpercent = 0
        for numcurr, curr in enumerate(cnorm):
            if curr[t] < 0:
                percent = int(abs(curr[t]) * (resy))
                imneg[lastpercent : lastpercent + percent, t] = numcurr
                lastpercent = lastpercent + percent
    im0 = vstack((impos, imneg))
    # CURRENTSCAPE CALCULATION ENDS HERE.

    # PLOT CURRENTSCAPE
    # adjust_left = None
    # adjust_right = 0.82
    # adjust_top = None
    # adjust_bottom = None
    # curr_names = [
    #     "Ih",
    #     "Ca_HVA2",
    #     "Ca_LVAst",
    #     "SK_E2",
    #     "SKv3_1",
    #     "K_Pst",
    #     "K_Tst",
    #     "NaTg",
    # ]
    # voltage_ticks = [-50, -20]
    # voltage_ylim = (-90, 30)
    # voltage_units = "[mV]"
    # voltage_color = "black"
    # current_ticks = [5, 50, 500]
    # current_ylim = (0.01, 1500)
    # current_units = "[pA]"
    # current_color = "black"
    # in_label = "inward %"
    # out_label = "outward %"

    # labelpad = 1
    # elcolormap = "Set1"
    # title = "XX pA"
    # figsize = (3, 4)
    # textsize = 6
    # legendsize = 6
    # titlesize = 12

    # set voltage ticks if not forced by config
    if not c["voltage"]["ticks"]:
        low_tick = int(voltage[0])  # round down
        high_tick = round(low_tick + 30, -1)
        c["voltage"]["ticks"] = [low_tick, high_tick]
    # set text size
    rcParams["axes.labelsize"] = c["textsize"]
    rcParams["ytick.labelsize"] = c["textsize"]
    rcParams["legend.fontsize"] = c["legendsize"]
    # remove legend handles
    rcParams["legend.handletextpad"] = 0
    rcParams["legend.labelspacing"] = 0
    rcParams["legend.handlelength"] = 0

    # print(c)
    resy = c["currentscape"]["y_resolution"]

    # START PLOT
    fig = figure(figsize=c["figsize"])
    if c["title"]:
        fig.suptitle(c["title"], fontsize=c["titlesize"])

    # PLOT VOLTAGE TRACE
    xmax = len(voltage)
    ax = subplot2grid((7, 1), (0, 0), rowspan=2)
    t = arange(0, len(voltage))
    plot(t, voltage, color=c["voltage"]["color"], lw=1.0)
    plot(
        t, ones(len(t)) * c["voltage"]["ticks"][0], ls="dashed", color="black", lw=0.75
    )
    vlines(1, c["voltage"]["ticks"][0], c["voltage"]["ticks"][-1], lw=1, color="black")
    ylim(c["voltage"]["ylim"])
    xlim(0, xmax)
    if c["show"]["labels"]:
        ax.set_ylabel(c["voltage"]["units"], labelpad=c["labelpad"])
    if c["show"]["ticklabels"]:
        ax.set_yticks(c["voltage"]["ticks"])
    remove_ticks_and_frame(ax)

    # PLOT TOTAL INWARD CURRENT IN LOG SCALE
    ax = subplot2grid((7, 1), (2, 0), rowspan=1)
    fill_between(arange(len((npPD))), (npPD), color=c["current"]["color"])
    for tick in c["current"]["ticks"]:
        plot(tick * ones(len(nnPD)), color="black", ls=":", lw=1)
    yscale("log")
    ylim(c["current"]["ylim"])
    xlim(0, xmax)
    if c["show"]["labels"]:
        ax.set_ylabel("+" + c["current"]["units"], labelpad=c["labelpad"])
    if c["show"]["ticklabels"]:
        ax.set_yticks(c["current"]["ticks"])
        ax.get_yaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())
    remove_ticks_and_frame(ax)

    # PLOT CURRENT SHARES
    ax = subplot2grid((7, 1), (3, 0), rowspan=3)
    im = imshow(
        im0[::1, ::1],
        interpolation="nearest",
        aspect="auto",
        cmap=c["currentscape"]["cmap"],
    )
    ylim(2 * resy, 0)
    plot(resy * ones(len(npPD)), color="black", lw=2)
    plt.gca().xaxis.set_major_locator(plt.NullLocator())
    plt.gca().yaxis.set_major_locator(plt.NullLocator())
    xlim(0, xmax)
    clim(0, len(cnorm))  # not sure about the vmax
    if c["show"]["labels"]:
        # cheat to set two labels at the right places for 1 axis
        set_label(
            ax,
            x=0,
            y=3 * resy / 2,
            label=c["currentscape"]["in_label"],
            textsize=c["textsize"],
        )
        set_label(
            ax,
            x=0,
            y=resy / 2,
            label=c["currentscape"]["out_label"],
            textsize=c["textsize"],
        )
    # legend
    if c["show"]["legend"]:
        set_legend(ax, im, c["current"]["names"])

    # PLOT TOTAL OUTWARD CURRENT IN LOG SCALE
    ax = subplot2grid((7, 1), (6, 0), rowspan=1)
    fill_between(arange(len((nnPD))), (nnPD), color=c["current"]["color"])
    for tick in c["current"]["ticks"]:
        plot(tick * ones(len(nnPD)), color="black", ls=":", lw=1)
    yscale("log")
    ylim(c["current"]["ylim"][1], c["current"]["ylim"][0])  # inverse axis
    xlim(0, xmax)
    if c["show"]["labels"]:
        ax.set_ylabel("-" + c["current"]["units"], labelpad=c["labelpad"])
    if c["show"]["ticklabels"]:
        ax.set_yticks(c["current"]["ticks"])
        ax.get_yaxis().set_major_formatter(matplotlib.ticker.ScalarFormatter())
    remove_ticks_and_frame(ax)

    adjust(
        c["adjust"]["left"],
        c["adjust"]["right"],
        c["adjust"]["top"],
        c["adjust"]["bottom"],
    )

    if c["output"]["savefig"]:
        if not os.path.exists(c["output"]["dir"]):
            os.makedirs(c["output"]["dir"])
        out_path = (
            os.path.join(c["output"]["dir"], c["output"]["fname"])
            + "."
            + c["output"]["extension"]
        )
        fig.savefig(
            out_path, dpi=c["output"]["dpi"], transparent=c["output"]["transparent"]
        )

    return fig
